trinity/protocol/wit/db.py

```diff
@@ -0,0 +1,61 @@
+from abc import ABC, abstractmethod
+
+from typing import Tuple
+
+from eth_typing import Hash32
+
+from eth.abc import AtomicDatabaseAPI
+
+import rlp
+
+from trinity._utils.async_dispatch import async_method
+
+
+class AsyncWitnessDataBaseAPI(ABC):
+    @abstractmethod
+    async def coro_get_witness_hashes(self, block_hash: Hash32) -> Tuple[Hash32, ...]:
+        ...
+
+    @abstractmethod
+    async def coro_persist_witness_hashes(
+        self,
+        block_hash: Hash32,
+        witness_hashes: Tuple[Hash32, ...],
+    ) -> None:
+        ...
+
+
+class AsyncWitnessDB(AsyncWitnessDataBaseAPI):
+    _recent_blocks_with_witnesses_lookup_key = b'recent-blocks-with-witness-hashes'
+    _max_witness_history = 256
+
+    def __init__(self, db: AtomicDatabaseAPI) -> None:
+        self.db = db
+
+    def _make_block_witness_hashes_lookup_key(self, block_hash: Hash32) -> bytes:
+        return b'block-witness-hashes:%s' % block_hash
+
+    def _get_recent_blocks_with_witnesses(self) -> Tuple[Hash32, ...]:
+        return tuple(rlp.decode(self.db[self._recent_blocks_with_witnesses_lookup_key]))
+
+    def get_witness_hashes(self, block_hash: Hash32) -> Tuple[Hash32, ...]:
+        return tuple(rlp.decode(self.db[self._make_block_witness_hashes_lookup_key(block_hash)]))
+
+    def persist_witness_hashes(
+            self, block_hash: Hash32, witness_hashes: Tuple[Hash32, ...]) -> None:
+        try:
+            recent_blocks_with_witnesses = list(self._get_recent_blocks_with_witnesses())
+        except KeyError:
+            recent_blocks_with_witnesses = []
+
+        if len(recent_blocks_with_witnesses) == self._max_witness_history:
```

Comment:
If there was a bug that caused the witness list to get too long, this check would stop removing old witnesses.

One option is to do this:
```suggestion
        while len(recent_blocks_with_witnesses) >= self._max_witness_history:
```